Cubical/Homotopy/Group/Pi4S3/BrunerieIso.agda

```diff
@@ -0,0 +1,526 @@
+
+-- We instantiate the sequence
+-- π₃ P → π₃ S² → π₃ coFib-fold∘W∙ → π₂ P
+P→S²→Pushout :
+  Exact4 (πGr 2 P) (πGr 2 (S₊∙ 2)) (πGr 2 coFib-fold∘W∙) (πGr 1 P)
+        (LESinst.fib→A 2)
+        (LESinst.A→B 2)
+        (LESinst.B→fib 1)
+Exact4.ImG→H⊂KerH→L P→S²→Pushout = LESinst.Im-fib→A⊂Ker-A→B 2
+Exact4.KerH→L⊂ImG→H P→S²→Pushout = LESinst.Ker-A→B⊂Im-fib→A 2
+Exact4.ImH→L⊂KerL→R P→S²→Pushout = LESinst.Im-A→B⊂Ker-B→fib 1
+Exact4.KerL→R⊂ImH→L P→S²→Pushout = LESinst.Ker-B→fib⊂Im-A→B 1
+
+-- The goal now is to rewrite it as
+-- π₃ S³ → π₃ S² → π₃ coFib-fold∘W∙ → Unit using the
+-- "functions from spheres"-definition of πₙ.
+-- Here, the first map is the one induced by fold∘W. We do this by:
+-- (1) showing that π₂ P is trivial
+-- (2) extending the sequence by appending surjections
+-- π₃ S³ → π₃ TotalPushoutPath×∙ → π₃ P on the left.
+-- (3) proving that this new composition is indeed the appropriate map
+
+-- Step 1: π₂ P is trivial
+π₂P≅0 : GroupEquiv (πGr 1 P) UnitGr
+π₂P≅0 = compGroupEquiv (πIso (isoToEquiv fiberinr'Iso , refl) 1)
+         (GroupIso→GroupEquiv
+           (contrGroupIsoUnit
+             (isOfHLevelRetractFromIso 0 (invIso iso₂) isContrπ₂S³)))
+  where
+  iso₁ : Iso (hLevelTrunc 4 (S₊ 3))
+             (hLevelTrunc 4 (Σ (Unit × S₊ 2) PushoutPath×))
+  iso₁ = connectedTruncIso 4 S³→TotalPushoutPath× isConnected-toPullback
+
+  iso₂ : Iso (π 2 (hLevelTrunc∙ 4 (S₊∙ 3)))
+              (π 2 TotalPushoutPath×∙)
+  iso₂ =
+    (compIso (setTruncIso
+      (equivToIso (_ ,
+        (isEquivΩ^→ 2 (fun iso₁ , refl) (isoToIsEquiv iso₁)))))
+     (invIso (πTruncIso 2)))
+
+  isContrπ₂S³ : isContr (π 2 (hLevelTrunc∙ 4 (S₊∙ 3)))
+  isContrπ₂S³ =
+    subst (λ x → isContr (π 2 x))
+      (λ i → ((sym (isContr→≡Unit (sphereConnected 3))) i)
+            , transp (λ j → isContr→≡Unit
+              (sphereConnected 3) (~ i ∧ j)) i ∣ north ∣)
+      (∣ refl ∣₂ , sElim (λ _ → isSetPathImplicit)
+                  λ p → cong ∣_∣₂
+                          (isProp→isSet
+                           (isOfHLevelPath 1 isPropUnit _ _) _ _ _ p))
+
+-- Step 2. We transform our original sequence to one for the
+-- the "maps from spheres" definition of πₙ and where π₂ P is
+-- replaced by the trivial group:
+-- π₃ P → π₃ S² → π₃ coFib-fold∘W∙ → 0
+P→S²→Pushout→P' :
+  Exact4 (π'Gr 2 P) (π'Gr 2 (S₊∙ 2)) (π'Gr 2 coFib-fold∘W∙) UnitGr
+         (π'∘∙Hom 2 (fst , refl))
+         (π'∘∙Hom 2 inr∙)
+         (→UnitHom _)
+P→S²→Pushout→P' =
+  transportExact4
+  (sym (GroupPath _ _ .fst ((GroupIso→GroupEquiv (π'Gr≅πGr 2 P)))))
+  (sym (GroupPath _ _ .fst ((GroupIso→GroupEquiv (π'Gr≅πGr 2 (S₊∙ 2))))))
+  (sym (GroupPath _ _ .fst ((GroupIso→GroupEquiv (π'Gr≅πGr 2 coFib-fold∘W∙)))))
+  (sym (GroupPath _ _ .fst π₂P≅0))
+  _ _ _ _ _
+  P→S²→Pushout
+  (toPathP (Σ≡Prop (λ _ → isPropIsGroupHom _ _)
+    (fromPathP λ i → fst (π∘∙fib→A-PathP 2 inr∙ i))))
+  ((toPathP (Σ≡Prop (λ _ → isPropIsGroupHom _ _)
+    (fromPathP λ i → fst (π∘∙A→B-PathP 2 inr∙ i)))))
```

Comment:
This `toPathP` composed with `fromPathP` looks kind of funny, but maybe there's not nicer way to do it?